rir/src/ir/Compiler.cpp

```diff
@@ -339,247 +339,83 @@ bool compileSpecialCall(Context& ctx, SEXP ast, SEXP fun, SEXP args_) {
-        //                       rewrite);
-        //}
-
-        // cs << BC::stvar(target)
-        //   << BC::invisible();
-
-        // return true;
+        // Find all parts of the lhs
+        SEXP target = nullptr;
+        l = lhs;
+        std::vector<SEXP> lhsParts;
+        while (!target) {
+            Match(l) {
+                Case(LANGSXP, fun, args) {
+                    assert(TYPEOF(fun) == SYMSXP);
+                    lhsParts.push_back(l);
+                    l = CAR(args);
+                }
+                Case(SYMSXP) {
+                    lhsParts.push_back(l);
+                    target = l;
+                }
+                Case(STRSXP) {
+                    assert(Rf_length(l) == 1);
+                    target = Rf_install(CHAR(STRING_ELT(l, 0)));
+                    lhsParts.push_back(target);
+                }
+                Else({
+                    errorcall(ast,
+                              "invalid (do_set) left-hand side to assignment");
+                })
+            }
+        }
+
+        // TODO: There is some issue with supper assing and probably [[, needs
+        // to investigate more...
+        if (superAssign)
+            return false;
+
+        // 3) Special case [ and [[
+        if (lhsParts.size() == 2) {
+            RList g(lhsParts[0]);
+            if (g.length() == 3) {
+                SEXP fun2 = *g.begin();
+                auto idx = g.begin() + 2;
+                if (*idx != R_DotsSymbol && *idx != R_MissingArg &&
+                    !idx.hasTag()) {
+                    if (fun2 == symbol::DoubleBracket ||
+                        fun2 == symbol::Bracket) {
+
+                        cs << BC::guardNamePrimitive(fun);
+
+                        // First rhs (assign is right-associative)
+                        compileExpr(ctx, rhs);
+                        // Keep a copy of rhs since its the result of this
+                        // expression
+                        cs << BC::dup() << BC::setShared();
+
+                        // Now load index and target
+                        cs << (superAssign ? BC::ldvarSuper(target)
+                                           : BC::ldvar(target));
+                        if (superAssign)
+                            cs << BC::setShared();
+                        compileExpr(ctx, *idx);
+
+                        // do the thing
+                        if (fun2 == symbol::DoubleBracket)
+                            cs << BC::subassign2();
+                        else
+                            cs << BC::subassign1();
+                        cs.addSrc(ast);
+
+                        // store the result as "target"
+                        cs << (superAssign ? BC::stvarSuper(target)
+                                           : BC::stvar(target));
+
+                        cs << BC::invisible();
+                        return true;
+                    }
+                }
```

Comment:
The rest is deleted intentionally?

Reply:
yeah, it's bitrotted anyway and we can loot into the git history if we ever want.